Added Convert syntax errors to violations.

laurentmuller opened the pull request:
This is a implementation for the issue #131.

laurentmuller committed:
Merge pull request #1 from friendsoftwig/master

Updated from upstream.

laurentmuller committed:
Added Convert syntax errors to violations.

OwlyCode commented:
Hey nice addition!

I'm convinced this should be turned on by default to avoid the tool to completely break on syntax errors like it does at the moment. This also would avoid to introduce a new option to the CLI.

Also, can you add a test for that?

laurentmuller commented:
OK,

I will turn option on by default and create 2 tests for an invalid template:

- Option on: Must not throw an exception

- Option off: Must throw a `SyntaxError`.

laurentmuller committed:
Updated LintCommand with default syntax error conversion.

Added unit tests and data for syntax error argument.
Updated the optional argument.

laurentmuller commented:
I have updated the LintCommand. The optional argument has been changed to `--error`. Please see the [Options with optional arguments](https://symfony.com/doc/current/console/input.html#options-with-optional-arguments). So, if the argument is `false` or missing, the error is not throw.

oallain commented on src/Console/LintCommand.php:
```diff
@@ -14,6 +15,7 @@
 use function class_exists;
 use function sprintf;
 
+
```
```suggestion

```

oallain commented on src/Console/LintCommand.php:
```diff
@@ -66,14 +69,28 @@ public function execute(InputInterface $input, OutputInterface $output)
+                file_get_contents($realPath),
+                $realPath,
+                str_replace(realpath($path), rtrim($path, '/'), $realPath)
+                );
```
```suggestion
            );
```

oallain commented on src/Console/LintCommand.php:
```diff
@@ -124,4 +141,16 @@ private function getSeverityLimit(InputInterface $input): int
+     * @param \Symfony\Component\Console\Input\InputInterface $input
+     * @return bool
+     */
+    private function hasEmptyConvertOption(InputInterface $input)
```
```suggestion
    private function hasEmptyConvertOption(InputInterface $input): bool
```

Ciloe replied:
This function can be deleted. This is not used

laurentmuller commented:
Hello,
I can understand why the tests failed. On my computer all the tests pass.
Any idea?

laurentmuller commented:
![tests_twigcs](https://user-images.githubusercontent.com/4330059/76565134-40a51500-64aa-11ea-92c9-55531d9dbb0f.png)

oallain commented:
I think it's because php7.0 don't support nullable return.
Error is in this file
https://github.com/friendsoftwig/twigcs/blob/59baa5897d2c92b08221c1e74a311afc5ba018a6/src/TwigPort/SyntaxError.php#L37

We need to increase requirement to PHP 7.1+ ?

Ciloe commented on src/Console/LintCommand.php:
```diff
@@ -29,6 +31,7 @@ public function configure()
             ->addOption('severity', 's', InputOption::VALUE_REQUIRED, 'The maximum allowed error level.', 'warning')
             ->addOption('reporter', 'r', InputOption::VALUE_REQUIRED, 'The reporter to use.', 'console')
             ->addOption('display', 'd', InputOption::VALUE_REQUIRED, 'The violations to display, "' . self::DISPLAY_ALL . '" or "' . self::DISPLAY_BLOCKING . '".', self::DISPLAY_ALL)
+            ->addOption('error', 'e',  InputOption::VALUE_OPTIONAL, 'Throw syntax error when a template contains an invalid token.', false)
```
I think that syntax-error is more explicit

```suggestion
            ->addOption('syntax-error', 'e',  InputOption::VALUE_OPTIONAL, 'Throw syntax error when a template contains an invalid token.', false)
```

OwlyCode replied:
May I suggest `throw-syntax-errors`? Looks to me that we're ready to merge after that!

laurentmuller committed:
Updated SyntaxError to be compatible with PHP 7.0.

laurentmuller commented:
I have updated SyntaxError to be compatible with PHP 7.0.

oallain commented:
> I have updated SyntaxError to be compatible with PHP 7.0.

It fix your issue, but I think it's shame to support already PHP 7.0 ;)

laurentmuller committed:
Removed hasEmptyConvertOption function from LintCommand.

laurentmuller committed:
Implementation of PR #133.

OwlyCode committed:
Merge branch 'master' into master

OwlyCode commented:
@laurentmuller I fixed the conflicts and merged, the option has been renamed to `--throw-syntax-errors`

stof commented on src/Console/LintCommand.php:
```diff
@@ -29,6 +31,7 @@ public function configure()
             ->addOption('severity', 's', InputOption::VALUE_REQUIRED, 'The maximum allowed error level.', 'warning')
             ->addOption('reporter', 'r', InputOption::VALUE_REQUIRED, 'The reporter to use.', 'console')
             ->addOption('display', 'd', InputOption::VALUE_REQUIRED, 'The violations to display, "'.self::DISPLAY_ALL.'" or "'.self::DISPLAY_BLOCKING.'".', self::DISPLAY_ALL)
+            ->addOption('throw-syntax-error', 'e',  InputOption::VALUE_OPTIONAL, 'Throw syntax error when a template contains an invalid token.', false)
```
Why `VALUE_OPTIONAL` rather than a `VALUE_NONE` flag when the presence of the flag is what matters ?
